### PR TITLE
ENH: Allow to assign/create custom body_style and header_style property to an instance of ExcelFormatter used in method df.to_excel()

### LINKED ISSUE #52369: ENH: Allow to assign/create custom body_style and header_style property to an instance of ExcelFormatter used in method df.to_excel()
### Feature Type

- [ ] Adding new functionality to pandas

- [X] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

I would like to assign a basic custom style for DataFrame body when it is exported to Excel with the method  `pandas.DataFrame.to_excel()`.  I know that there are a vast number of alternatives using stylers but those use large libraries such as matplotlib to produce a simple css style, for example, if the user just needs to add borders or bold text or alignment.

Currently, the method `pandas.DataFrame.to_excel()` uses by default for any DataFrame a style with borders and bold for headers and indexes, and no style for the rest of the body.

 

### Feature Description

There are many alternatives. For me the shortest and easiest one with the less number of changes in the current code will be:

According to the source code of the current dev-version, the method `pandas.DataFrame.to_excel()` uses an instance of a class `ExcelFormatter` to format cells and then writes them to a file. This class is defined in file `pandas/io/formats/excel.py` and its method `.write()` calls the method `.get_formatted_cells()` and it in its turn calls the private method `._format_body()` and after some more calls the method `._generate_body()` is reached. This method formats series by series and by default uses a `style=None`  which is hard coded.

```python
# Final part of current `ExcelFormatter._generate_body()`
yield CssExcelCell(
        row=self.rowcounter + i,
        col=colidx + coloffset,
        val=val,
        style=None,
        css_styles=getattr(self.styler, "ctx", None),
        css_row=i,
        css_col=colidx,
        css_converter=self.style_converter,
        )
```

I would like to change that `None` default value to a new ExcelFormatter property similar to the current `header_style` used to format headers and indexes. That needs to add a new property and change the call to CssExcelCell in line ~ 882 in file `pandas/io/formats/excel.py` as I show in this shortcode

```python
# ExcelFormatter properties
@property
def header_style(self) -> dict[str, dict[str, str | bool]]:
    return {
        "font": {"bold": True},
        "borders": {
            "top": "thin",
            "right": "thin",
            "bottom": "thin",
            "left": "thin",
        },
        "alignment": {"horizontal": "center", "vertical": "top"},
    }
# -- >  NEW PROPERTY < ---
@property
def body_style(self) -> dict[str, dict[str, str | bool]]:
    return {
        "font": {"bold": False},
        "borders": {
            "top": "thin",
            "right": "thin",
            "bottom": "thin",
            "left": "thin",
        },
        "alignment": {"horizontal": "right", "vertical": "top"},
    }
# -- > END  NEW PROPERTY < ---

# Call in `ExcelFormatter._generate_body()` to ``CssExcelCell`
```python
yield CssExcelCell(
     # ...
     style=self.body_style
     # ...
    )
```

I don't know if there is a better way to set those properties in pandas, but for my purposes, I can build a new style as a `dict` in my code and assign my new styles to the properties or set to `None` before calling the constructor of ExcelFormatter in the method.

```python

pd.io.formats.excel.ExcelFormatter.header_style = style_header
pd.io.formats.excel.ExcelFormatter.body_style = style_body

df.to_excel(writer, sheet_name=sheet_name)

```
<hr>

A more sophisticated alternative will included new arguments with custom styles for the method ``pandas.DataFrame.to_excel()`  and changes in the way those styles are parsed by ExcelFormatter instances, but in the end maybe look similar to the shortest alternative that I describe above.

### Alternative Solutions

As an alternative users can use Stylers in DataFrame. This alternative requires matplotlib, jinja2.

### Additional Context

I already tested these small changes in my code in an environment with pandas (version 1.4.2) and xlsxwriter I would like to push those changes for a more recent version. But, in virtue that I don't know which tests are usually made by the developers' team related to complex DataFrames or performance, I decided to launch this request.

### PR DESCRIPTION
**Pull request by rmhowe425**
- [X] closes #52369 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by rmhowe425** [comment]
@rhshadrach I'm reading through the linked GitHub issue, as well as @attack68's comments. Do we want to use setter methods to declare values for `body_style` and `header_style`? 

Or are we  letting `Styler` be used to add and define all the custom style?

**Comment by attack68** [comment]
My preferred option as stated in the issues was to remove the formatting added by default when doing `df.to_excel()`, and make the user use `Styler` to control all of the output formatting.

However, given that might annoy a number of users I would suggest a compromise.

1) Adding `properties` and `setters` as per one of the issues.
2) Default these default properties to no styling later in a major release and document how a user can set these properties to retrieve the original behaviour (bolding  /bordering the header / index cells).

File: pandas/io/formats/excel.py [hunk 1]
```diff
@@ -577,19 +577,32 @@ def __init__(
         self.header = header
         self.merge_cells = merge_cells
         self.inf_rep = inf_rep
+        self._header_styledict[str, Any] = None
```
Comment (rhshadrach) [inline pandas/io/formats/excel.py:580]: Should this be `self._header_style: dict[str, Any] | None = None`

Comment (rhshadrach) [reply pandas/io/formats/excel.py:580]: What is the difference between defaulting to `None` and `{}` here?

File: pandas/io/formats/excel.py [hunk 2]
```diff
@@ -577,19 +577,32 @@ def __init__(
...
-            },
-            "alignment": {"horizontal": "center", "vertical": "top"},
-        }
+    def header_style(self):
+        return self._header_style
+
+    @header_style.setter
+    def header_style(self, val):
+        if not isinstance(val, dict):
+            return None
```
Comment (rhshadrach) [inline pandas/io/formats/excel.py:590]: setters shouldn't return a value

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: doc/source/user_guide/io.rst [hunk 3]
```diff
@@ -3880,6 +3880,19 @@ The look and feel of Excel worksheets created from pandas can be modified using
...
 * ``freeze_panes`` : A tuple of two integers representing the bottommost row and rightmost column to freeze. Each of these parameters is one-based, so (1, 1) will freeze the first row and first column (default ``None``).
 
+The styling and formatting of a worksheet's header and body can be easily customized using CSS by passing a dictionary object to
+``pd.io.formats.excel.ExcelFormatter.header_style`` and ``pd.io.formats.excel.ExcelFormatter.body_style``.
+
+.. code-block:: python
+
+    # Set Excel worksheet header and body style
+    >>> df = pd.DataFrame([{"A": 1, "B": 2, "C": 3}, {"A": 1, "B": 2, "C": 3}])
+    >>> pd.io.formats.excel.ExcelFormatter.header_style = {"font": {"bold": True},}
```
Comment (rhshadrach) [inline doc/source/user_guide/io.rst:3890]: I don't think this is doing what you might expect. In particular, this is not using getters/setters, but rather just overwriting a property.

```
class Foo:
    @property
    def foo(self):
        raise ValueError
        
    @foo.setter
    def foo(self, other):
        raise ValueError
        
print(Foo.foo)
# <property object at 0x7f5438093290>
Foo.foo = 5
print(Foo.foo)
# 5
```

**Review by rhshadrach — CHANGES_REQUESTED** [review]
This is overwriting a property on a class, replacing it with (presumably) a Python dictionary. More generally, it is relying on modifying global state but not by any standard approach (e.g. options). I _think_ it is also requiring users to access a class that was previously internal, further confusing the public/private nature of `pandas.io` that I think we should be trying to clear up.

For these reasons, I'm -1 on this approach in its current state. Is using an option for this appropriate?

File: doc/source/user_guide/io.rst (hunk 3 above)
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3890]: @rhshadrach @attack68 

What do we think would be the appropriate implementation for this request then? It seems like the [initial suggestion](https://github.com/pandas-dev/pandas/issues/52369) is not ideal.

Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3890]: @rhshadrach @attack68 

What do we think would be the appropriate implementation for this request then? It seems like the [initial suggestion](https://github.com/pandas-dev/pandas/issues/52369) is not ideal.

**Comment by rhshadrach** [comment]
@attack68 (from https://github.com/pandas-dev/pandas/pull/53973#issuecomment-1617487108)

> However, given that might annoy a number of users I would suggest a compromise.

Is it only the removing of the default style that you feel might annoy users? Or being forced into the Styler in order to add style? Or something else?

**Comment by attack68** [comment]
> Is it only the removing of the default style that you feel might annoy users? Or being forced into the Styler in order to add style? Or something else?

`to_excel` has been around for a while bolding the headers, perhaps users expect that by now? If we could deprecate that and use a keyword instead (like we do in to_latex and to_html) then I would prefer that, e.g. `bold_rows=False/True`. The method could re-route via the Styler implementation to do this (as it does for `to_latex`)

Since the Styler uses the underlying implementation of DataFrame.to_excel, it is a shame that an unstyled Styler ends up with bold headers by default because of the status quo.

> Pending a response to the above, I'm +1 on this approach. I think modifying global state should be avoided where ever possible, and this feature seems perfect for the Styler to me.

It does seem unnecessary to add this to the state variables of the excel writer when Styler can already do this by default and with greater flexibility. Perhaps documenting users to use the Styler is the better approach that making the code more complicated.

**Comment by rmhowe425** [comment]
@attack68 just to recap, add a parameter to `to_excel` that takes a boolean value representing whether the headers should be bolded. As you said, the value would propagate through the code so that the Styler implementation is responsible for formatting the output. 

For the formatting of the spreadsheet body, that would be controlled via `Styler` as well. 

@rhshadrach Are you good with this?

**Comment by attack68** [comment]
@rmhowe425 I just reviewed the current arguments for `DataFrame.to_excel`. It is actually *very* complicated to add a `bold_rows` argument that uses the Styler implementation  Not because this is complicated itself, but because you have to ensure all the other argument combinations that might used also result in the same excel file.  So let me save you the bother now - don't attempt it.

@rhshadrach - If we remove styling from `to_excel` and do not bold or create borders for any cells do we need a deprecation cycle for that, even though it isn't actually an option it is just created by default? As part of the deprecation we could write notes in the docs how to replicate similar features with the `Styler.to_excel` example?

**Comment by rhshadrach** [comment]
> If we remove styling from `to_excel` and do not bold or create borders for any cells do we need a deprecation cycle for that, even though it isn't actually an option it is just created by default?

I'm not sure I would even consider this a "breaking" change; I don't believe it could ever break any workflow. Even if it is considered breaking, I would lean toward adding this to the list of breaking changes in 3.0 in https://github.com/pandas-dev/pandas/issues/44823.

**Comment by rmhowe425** [comment]
@rhshadrach I'll plan on updating my implementation so that spreadsheets have no formatting by default, and I'll update documentation so that we're telling users to use `Styler` for formatting.

This sounds like the changes that we agree on?

**Comment by rhshadrach** [comment]
> This sounds like the changes that we agree on?

This is what I've suggested, but I think we need to make sure there is a consensus among those involved before saying "we agree".

**Comment by attack68** [comment]
> * ENH: Allow to assign/create custom body_style and header_style property to an instance of ExcelFormatter used in method df.to_excel() #52369

I have created an issue to highlight this to move forward: #54154 54154

**Comment by rmhowe425** [comment]
Closing PR as work is continuing on #54154

**Pull request closed without merging**